Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,390 @@
+    else{
+      if(isDark(DarkQuarkIDs_,dau)){
+        firstQdM.insert(dau);
+        if(secondDM == false){
+          firstQdM1 = dau;
+          secondDM = true;
+        }
+        else firstQdM2 = dau;
+      }
+      else if(isDark(SMQuarkIDs_,dau)){
+        firstQsM.insert(dau);
+        if(secondSM == false){
+          firstQsM1 = dau;
+          secondSM = true;
+        }
+        else firstQsM2 = dau;
+      }
+    }
+  }
+}
+
+void HiddenSectorProducer::firstDark(CandPtr part, CandSet& firstMd, CandSet& firstQd, CandSet& firstGd, CandSet& firstQdM, CandSet& firstQsM, CandPtr& firstQdM1, CandPtr& firstQdM2, CandPtr& firstQsM1, CandPtr& firstQsM2, bool& secondDM, bool& secondSM) const {
+  if(isDark(DarkFirstIDs_,part)){
+    CandPtr parent = part->mother(0);
+    if(isDark(DarkFirstIDs_,parent)) firstDark(parent, firstMd, firstQd, firstGd, firstQdM, firstQsM, firstQdM1, firstQdM2, firstQsM1, firstQsM2, secondDM, secondSM);
+    else{
+      //SM parent of first dark particles: fill first two vectors here, use mediators to fill second two vectors
+      for(unsigned i = 0; i < part->numberOfDaughters(); i++){
+        CandPtr dau = part->daughter(i);
+        if (isDark(DarkMediatorIDs_,dau)){
+          firstMd.insert(dau);
+          medDecay(dau,firstQdM,firstQsM,firstQdM1,firstQdM2,firstQsM1,firstQsM2,secondDM,secondSM);
+        }
+        else if (isDark(DarkQuarkIDs_,dau)) firstQd.insert(dau);
+        else if (isDark(DarkGluonIDs_,dau)) firstGd.insert(dau);
+      }
+    }
+  }
+}
+
+int HiddenSectorProducer::checkLast(const reco::GenJet& jet, CandSet stableDs, int value, double& frac) const {
+  //compare jet constituents to set of last particles
+  //also compute dark pt fraction in same loop
+  bool match = false;
+  LorentzVector p4;
+  LorentzVector totPt = jet.p4();
+  for(unsigned i = 0; i < jet.numberOfDaughters(); ++i){
+    CandPtr dau = jet.daughter(i);
+    if(isAncestor(DarkHadronIDs_,dau)){
+        match = true;
+        p4 += dau->p4();
+    }
+  }
+  for(const auto& part : stableDs){
+     if(reco::deltaR(jet, *part) < coneSize_)
+     {
+       p4 += part->p4();
+       totPt += part->p4();
+     }
+  }
+  frac = p4.pt()/totPt.pt();
+  return match ? value : 0;
+}
+
+int HiddenSectorProducer::checkFirst(const reco::GenJet& jet, const CandSet& firstP, int value) const {
+  //compare first particles using jet cone (can't compare constituents' parents because all dark hadrons descend from all first dark particles, strong force problems)
+  for(const auto& part : firstP){
+    if(reco::deltaR(jet, *part) < coneSize_) return value;
+  }
+  return 0;
+}
+
+// match particles from mediator to jets
+void HiddenSectorProducer::matchPFMtoJet(CandPtr& part, std::vector<LorentzVector>& matchedJets, const reco::GenJet& jet, int& nPartPerJet, int& t_MT2JetID, const int& mt2ID) const {
+  if(reco::deltaR(jet,*part) < coneSize_ and jet.pt() > 100)
+  {
+    matchedJets.push_back(jet.p4());
```

Comment:
avoid copying four-vectors unnecessarily: just use a vector of `reco::GenJet*` (then access `.p4()` when needed)